library/std/src/Std/DurrHoyerLibrary.qs

```diff
@@ -0,0 +1,207 @@
+            X(inputQubits[i]); // Flip qubits that should be zero in the threshold
+        }
+    }
+
+    // Controlled-Z gate to flip the phase of the state if the element is less than the threshold
+    Controlled Z(inputQubits, auxQubit);
+
+    // Undo the X operations to revert qubits
+    for i in 0..Length(thresholdBits) - 1 {
+        if (thresholdBits[i] == One) {
+            X(inputQubits[i]);
+        }
+    }
+}
+
+// Diffusion operator (Grover's diffusion)
+operation DiffusionOperator(qubits : Qubit[]) : Unit {
+    ApplyToEach(H, qubits);
+    ApplyToEach(X, qubits);
+    Controlled Z(qubits[0..Length(qubits) - 2], qubits[Length(qubits) - 1]);
+    ApplyToEach(X, qubits);
+    ApplyToEach(H, qubits);
+}
+
+// Grover iteration with the oracle and diffusion operator for min
+operation GroverIterationMin(threshold : Int, inputQubits : Qubit[], auxQubit : Qubit) : Unit {
+    OracleLessThan(threshold, inputQubits, auxQubit);
+    DiffusionOperator(inputQubits);
+}
+
+// Grover iteration with the oracle and diffusion operator for max
+operation GroverIterationMax(threshold : Int, inputQubits : Qubit[], auxQubit : Qubit) : Unit {
+    OracleMoreThan(threshold, inputQubits, auxQubit);
+    DiffusionOperator(inputQubits);
+}
+
+// Dürr-Høyer for finding min or max algorithm
+operation DurrHoyerAlgorithm(list : Int[], nQubits : Int, type : String, candidate : Int, lengthList : Int) : Int {
+    use inputQubits = Qubit[nQubits] {
+        use auxQubit = Qubit() {
+            // Create a superposition of all states
+            ApplyToEach(H, inputQubits);
+
+            // Continue Grover search until no better candidate is found
+            mutable betterCandidateFound = true;
+            mutable iterationCount = 1; // Track the iteration count manually
+            mutable optimalIterations = 5;
+            mutable validIterations = 0;
+
+            while (validIterations < optimalIterations) {
+                set betterCandidateFound = false;
+
+                // Calculate M: the number of elements smaller than the current candidate (for min)
+                let M : Int = CountElements(list, list[candidate], type);
+
+                // If there are no more elements smaller/larger, return the candidate
+                if (M == 0) {
+                    Message("No more elements to compare, search complete.");
+                    ResetAll(inputQubits + [auxQubit]);  // Ensure qubits are reset before returning
+                    return candidate;
+                }
+
+                // Calculate the optimal number of Grover iterations
+                let N : Int = Length(list);
+                let optimalIterations : Int = Round((PI() / 4.0) * Sqrt(IntAsDouble(N) / IntAsDouble(M)));
+
+                // Perform Grover iterations for min or max
+                for i in 1..optimalIterations {
+                    if (type == "min") {
```

Comment:
Re: Similarly here

Possibly rewrite this:

let groverIteration = (type=="min") ? GroverIterationMin | GroverIterationMax;
groverIteration(list[candidate], inputQubits, auxQubit)